Add support for collating sequences on indexed files

ddeclerck opened the pull request:
This PR adds support for COLLATING SEQUENCE on indexed files, as mentionned in [https://sourceforge.net/p/gnucobol/feature-requests/459/](https://sourceforge.net/p/gnucobol/feature-requests/459/).

Note this has. only been done for the BDB backend - that's what our client needs.

ddeclerck commented:
Hmm, that broke the NIST testsuite.
Investigating further, it seems we should in fact NOT apply the program collating sequence to indexed files that do not have an explicit collating sequence (as opposed to SORT/MERGE files).

See 12.3.5.3 in ISO/IEC 1989:2014 :
```quote
11) The alphanumeric program collating sequence and national program collating sequence are used to determine
the truth value of any alphanumeric comparisons and national comparisons, respectively, that are:
a)Explicitly specified in relation conditions.
b)Explicitly specified in condition-name conditions.
c)Implicitly specified by the presence of a CONTROL clause in a report description entry.
When alphabet-name-1 or alphabet-name-2, or both, is associated with a locale, locale category LC_COLLATE
is used to carry out these comparisons.

12) The alphanumeric program collating sequence and national program collating sequence explicitly or implicitly
established by the OBJECT-COMPUTER paragraph are effective with the initial state of the runtime modules
to which they apply. If alphabet-name-1 or alphabet-name-2 references a locale, the associated collating
sequence is defined by category LC_COLLATE in the specific locale associated with that alphabet-name or, if
none, in the locale current at the time the collating sequence is used at runtime

13) The alphanumeric program collating sequence and national program collating sequence are applied to
alphanumeric and national sort or merge keys, respectively, unless the sort or merge collating sequence has
been modified by execution of a SET statement or a COLLATING SEQUENCE phrase is specified in the
respective SORT or MERGE statement.```

GitMensch commented:
Found the answer in 12.4.4.3:

> If no COLLATING SEQUENCE clause is specified:
a) the  collating  sequence  for  alphanumeric  record  keys,  both  primary  and  alternate,  is  the  native alphanumeric collating sequence;
b) the collating sequence for national record keys, both primary and alternate, is the native national collating sequence.

So this is by default "native", not "program collating".
Please re-adjust, sorry if I was misleading here.

BUT: before adjusting the code, please add a _minimal_ version of the failing NIST test (ideally written from scratch) to our internal testsuite; this should fail before your change-adjustment and pass afterwards.

(and note that this is also one of the fixed file attributes, which in theory should be checked - we don't do up to 4.x, but don't check that there either...).

ddeclerck commented:
I made the necessary changes. I also added a new `-fdefault-file-colseq` flags that works similarly to `-fdefault-colseq` but just for files.

As for the tests, I reworked them and included a dummy program collating sequence (taken from the failed NIST test), which was sufficient to cause the testsuite to fail priori to fixing the bug.

GitMensch commented:
That sounds all good!

Please check the test coverage of your changes, the codecov bot says:

> Attention: 188 lines in your changes are missing coverage. Please review.

188 sounds a bit much.

GitMensch commented on cobc/parser.y:
```diff
@@ -5752,7 +5776,6 @@ collating_sequence_clause:
 	check_repeated ("COLLATING", SYN_CLAUSE_3, &check_duplicate);
 	current_file->collating_sequence = alphanumeric_collation;
 	current_file->collating_sequence_n = national_collation;
-	CB_PENDING ("FILE COLLATING SEQUENCE");
```
This should be changed to `CB_UNFINISHED` - it only works for BDB (LMDB has this function internal only, D-ISAM and likely C-ISAM have no option for passing that from the outside (D-ISAM has `ISCUSTKEYS`, but that is only available with licensed versions built from source with custom compare function added), V*ISAM has the extension `issetcollate()` - but so far the passed collation is stored and then never used).

Please also add missing pieces for `collating_sequence_keys` - the `#ifdef 0` can be removed from `cob_file_key` so that you have the matching field (it is only allocated in libcob); the code in fileio.c to set the key into app_data would first check the key, then the file. The most work here is in the compiler and also those should be not that much..

ddeclerck commented:
> That sounds all good!
> 
> Please check the test coverage of your changes, the codecov bot says:
> 
> > Attention: 188 lines in your changes are missing coverage. Please review.
> 
> 188 sounds a bit much.

Fixed ! Just had to resynchronize with upstream.
The remaining uncovered lines are either benign or were not covered anyways.

GitMensch commented on libcob/coblocal.h:
```diff
@@ -570,6 +570,9 @@ cob_max_int (const int x, const int y)
 	return y;
 }
 
+COB_HIDDEN int		common_cmps	(const unsigned char *, const unsigned char *,
```
so far all `COB_HIDDEN` have a `cob_` prefix. please refactor this to be identical

GitMensch commented on libcob/common.h:
```diff
@@ -1395,7 +1395,7 @@ typedef struct __cob_file {
 	void			*file;			/* File specific pointer */
 	void			*linorkeyptr;		/* LINAGE or SPLIT KEY */
-	const unsigned char	*sort_collating;	/* SORT collating / CODE-SET (used for RE-/WRITE) */
+	const unsigned char	*sort_collating;	/* SORT collating */
```
please adjust comment as this is now also the FILE collation for INDEXED files

ddeclerck replied:
Actually canceled this change, since the file collation is now directly stored in the keys.

GitMensch commented on libcob/common.h:
```diff
@@ -1432,6 +1432,7 @@ typedef struct __cob_file {
 	char				*nxt_filename;	/* Next position in org_filename */
 	unsigned int		flag_is_concat:1;	/* SEQUENTIAL concatenated file names */
 
+	const unsigned char* code_set_write;	/* CODE-SET conversion for RE-/WRITE) */
```
If you want to do this splitting - which seems reasonable - then you need to check the module version and "move" the field for GC 3.2 generated modules, as those will have used the field above. Also you'd have to move it at the end of the `cob_file`, as hinted in the doc-comment for `__cob_file`.

Note: If you _always_ place the key collation into _each_ `cob_key`, then you don't need to have the file-collation in `cob_file` at all (if set then those will never be empty).

GitMensch commented on libcob/fileio.c:
```diff
@@ -5429,6 +5464,7 @@ indexed_read_next (cob_file *f, const int read_opts)
 		} else {
 			p->key.size = (cob_dbtsize_t) bdb_keylen(f,p->key_index);
 			p->key.data = p->last_readkey[p->key_index];
+			bdb_setkeycol (f, p->key_index);
```
when placing this into each `cob_key` you can direclty do `p->key.app_data = f->keys[p->key_index]->collation` here

GitMensch commented on tests/testsuite.src/run_file.at:
```diff
@@ -12429,6 +12429,241 @@ AT_CHECK([diff reference prog.out], [0], [], [])
+AT_SETUP([INDEXED file manipulation under ASCII/EBCDIC collation])
+AT_KEYWORDS([runfile WRITE DELETE READ EBCDIC])
+
+AT_SKIP_IF([test "$COB_HAS_ISAM" = "no"])
```
Change to `AT_SKIP_IF([test "$COB_HAS_ISAM" != "db"])` and add a comment that this is only included for the BDB backend so far.

GitMensch reviewed (changes requested):
There are some general changes (should not be too much work) and one change to `cob_key` that actually reduces the amount of changes (here the important part again):

> Note: If you _always_ place the key collation into _each_ `cob_key`, then you don't need to have the file-collation in `cob_file` at all (if set then those will never be empty).**

I'd like to also have support for key collation - as the runtime already handles that (just specifying always the same collation), which is possibly a bit more work, if you feel that's too much, then this part could be postponed.

ddeclerck commented on libcob/fileio.c:
```diff
@@ -4150,6 +4164,23 @@ indexed_file_delete (cob_file *f, const char *filename)
+bdb_bt_compare(DB *db, const DBT *k1, const DBT *k2)
+{
+	int res = 0;
+	const unsigned char *col = (const unsigned char *)k1->app_data;
```
Done using a pair of macros, conditionally defined depending on BDB version.

ddeclerck commented:
Should be good. Though I left the key collating sequence for later (although I enabled the field to store the file collating sequence).

GitMensch commented on NEWS:
```diff
@@ -28,6 +28,8 @@ NEWS - user visible changes				-*- outline -*-
    build system do not correctly work together to locate files from
    diagnostic output
 
+** New option -fdefault-file-colseq to specify default file collating sequence
```
add  the file COLLATING SEQUENCE in general under "new features" (with the note "currently only with the BDB backend)

GitMensch commented on libcob/fileio.c:
```diff
@@ -757,6 +767,22 @@ bdb_savekey (cob_file *f, unsigned char *keyarea, unsigned char *record, int idx
+	/* We only apply a collating sequence if the key is alphanumeric */
+	if (COB_FIELD_IS_ANY_ALNUM(fld) || COB_FIELD_IS_NUMDISP(fld)) {
+		DBT_SET_APP_DATA(&p->key, (void *)f->keys[idx].collating_sequence);
+	/* TODO: this should be done for national, when available *//*
+	} else if (COB_FIELD_IS_NATIONAL(fld)) {
```
better use `#if 0 /* TODO ddd */` here, this leads to a better highlighting and less"that's a strange code here" analyzer warning

GitMensch commented on libcob/fileio.c:
```diff
@@ -757,6 +767,22 @@ bdb_savekey (cob_file *f, unsigned char *keyarea, unsigned char *record, int idx
 	return (int)f->keys[idx].field->size;
 }
 
+static void
+bdb_setkeycol (cob_file *f, int idx)
+{
+	struct indexed_file	*p = f->file;
+	const cob_field	*fld = f->keys[idx].field;
+	/* We only apply a collating sequence if the key is alphanumeric */
+	if (COB_FIELD_IS_ANY_ALNUM(fld) || COB_FIELD_IS_NUMDISP(fld)) {
+		DBT_SET_APP_DATA(&p->key, (void *)f->keys[idx].collating_sequence);
+	/* TODO: this should be done for national, when available *//*
+	} else if (COB_FIELD_IS_NATIONAL(fld)) {
+		DBT_SET_APP_DATA(&p->key, (void *)f->sort_collating_n);*/
+	} else {
+		DBT_SET_APP_DATA(&p->key, NULL);
+	}
+}
+
```
The complete check part here should not be done each time, but in the codegen, no?

GitMensch commented on libcob/fileio.c:
```diff
@@ -4150,6 +4179,23 @@ indexed_file_delete (cob_file *f, const char *filename)
+{
+	int res = 0;
+	const unsigned char *col = (const unsigned char *)DBT_GET_APP_DATA(k1);
+	if (col == NULL) {
```
Is there a reason to register `bdb_bt_compare` down there if `col == NULL`? I _think_ this check should be done outside.
Also: is it possible in our code that both DBT can have a different size? At least in the sort comparison they are always identical (and it would be a status 39 error if they wouldn't be here, no)?

ddeclerck replied:
I guess if no keys for the current file have a collation, then we don't need to register `bdb_bt_compare` indeed.

GitMensch replied:
I've thought this was registered per key; in this case if the key has no collation, then we don't need to register, and can get rid of the `col == NULL` check, otherwise it needs to stay in and we only don't register if there's no collation at all.

ddeclerck replied:
Oh, wait, you're right, the registration is indeed per key. So yeah, we can simplify the code further and register the comparison function only if the key has a collating sequence.

GitMensch replied:
That's part of the beauty of code-reviews - the code gets improved even before it gets live :-)
Wait, who reviews my code...

ddeclerck replied:
> Also: is it possible in our code that both DBT can have a different size? At least in the sort comparison they are always identical (and it would be a status 39 error if they wouldn't be here, no)?

Yeah, they should always be of identical length in our context (and I made a quick check by adding an assert and running both NIST and the testsuite with no error). Anyway I prefered to handle the case anyways (the comparison function is rather "low-level" and does not really care about our use case).

GitMensch reviewed (changes requested):
looks like we're nearly done; I think this can be checked in next week (and _maybe_ there is a key collation available until then, too)

GitMensch commented:
Note: I do have a bigger refactoring of the BDB code which I want to push start of next week, should I wait for this PR to be finished and merged before or would you possibly even prefer if I push the refactoring before?
(I just wanted to fix a bug, recognized that for doing so I should backport status 02 complex; then got from one thing to the next... and then did a refactoring, using the specific `f->keys[idx]` instead of using an index of a subfield and the same for the contained keydesc - the refactor got so big that I should push that separately in any case [the full code of course has a bug _somewhere_ and hides from me])

ddeclerck commented:
> Note: I do have a bigger refactoring of the BDB code which I want to push start of next week, should I wait for this PR to be finished and merged before or would you possibly even prefer if I push the refactoring before?

I plan to make the final changes to this PR on monday ; if that prevents you from merging, I don't mind rebasing my PR on your changes.

ddeclerck commented:
Made the last changes, though there might be another one to do.
I'm concerned about the memcmp here (fileio.c: 5679) :
```
	/* Check record key */
	bdb_setkey (f, 0);
	if (!p->last_key) {
		p->last_key = cob_malloc ((size_t)p->maxkeylen);
	} else
	if (f->access_mode == COB_ACCESS_SEQUENTIAL
	 && memcmp (p->last_key, p->key.data, (size_t)p->key.size) > 0) {
		return COB_STATUS_21_KEY_INVALID;
	}
	memcpy (p->last_key, p->key.data, (size_t)p->key.size);
```
All other memcmp's on keys are fin because they just test (in)equality with 0, but this one tests for `> 0`. I guess I should replace this with a collation-aware compare, right ?

GitMensch commented:
I have a strange feeling now... Can you please verify that:

* the testsuite has at least one entry with variable length keys (which test case is that?) for each SEARCH and INDEXED
* if missing please add it
* check if we do store the variable length in the key fields

I agree that a memcpy should only be used for testing equality or difference, otherwise either a numeric compare (if key field is numeric - the byte order could be non-native!) or a collation aware check needs to be used

ddeclerck commented:
> That was my feeling. Please address the numeric key sorting, if possible.

I see two ways to fix that :
- either modifiny the `bdb_bt_compare` function - but that would fix only for BDB...
- or always store integer numeric keys in big-endian order (not sure how to deal with IEE754...), but then if I actually translate the keys so thay end up stored in the right order, I might as well do that for alphanumeric keys, making this PR les relevant...

In any case, I believe this require some thinking before actually doing that, and I'm not sure this belongs to this PR (as it did not introduce that problem).

> I _think_ this should work (and hopefully be tested) in SEARCH already.

Tested, seems to work fine for SEARCH.

> As both can use the same comparison logic (just using a different set of collations), I think it's useful to test that.

Could you elaborate ?

> Testing variable length keys belong to this PR if previously possibly completely handled by BDB and now in our own search function.

I'm not sure to understand ; could you rephrase that ?

GitMensch commented:
> * either modifying the `bdb_bt_compare` function - but that would fix only for BDB...

that would be enough for now and is likely "the most portable" solution. But that creates a HUGE problem: if we do that, then the files need to be rebuilt :-/  Note that using the key type and use the appropriate COBOL sort function (if the field structure within ´->app_data` [if not done: store a complete cb_field structure in there], then you only need to duplicate that cob_field to then set the two separate fields' data) should directly work for all types. This is effectively what is done for `SEARCH` (and the internal comparison during `SORT`).

Otherwise we could have small-enough numbers be converted to big-endian binary and bigger ones to be converted to USAGE DISPLAY SIGN SEPARATE LEADING in `bdb_savekey`, that would also work - but again leads to all files needing a rebuild. As this repacking/unpacking would be done on any key access, this is likely slower than the approach above.

As LMDB does not provide an external compare function, the second solution seems to be the only one that could be applied there (but note that LMDB exists only in trunk and even then is marked as experimental and possibly will be dropped).

Note: with ISAM there are key types that handle this (including IEE754), but we currently only use `CHARTYPE`, so the same problem seems to exists there. For this part I'll contact Ron.

> In any case, I believe this require some thinking before actually doing that, and I'm not sure this belongs to this PR (as it did not introduce that problem).

Hm. Please add a test with expected failure and a comment about the underlying problem.
We could otherwise leave that general problem out of this PR, but if a key is purely numeric, there should be no "additional" conversion applied which makes the current scenario worse.
That brings us to the good thing: if you implement key collation within the cobc site (you already have the codegen and use of that in libcob) then you can place a check there for a numeric type and just skip generating the collation for that.

>> As both can use the same comparison logic (just using a different set of collations), I think it's useful to test that.
> 
> Could you elaborate ?

`SEARCH`/`SORT` already compares "as cob_field" for any numeric type; if the same is applied to the INDEXED comparisons (you need a `cob_field` in ´app_data` for that), then everything "just works".
But as we noted: this may be kept to a separate PR, _especially_ as we have to consider what to do with the existing files (likely we'd have to insert a marker into the file itself, likely by an embedded table, where we record "that's type 2, use different search algorithm"). This is definitely useful, but definitely "quite big" (and yes, should be kept separate).

>> Testing variable length keys belong to this PR if previously possibly completely handled by BDB and now in our own search function.
> 
> I'm not sure to understand ; could you rephrase that ?

BDB compares keys. those are setup from `cob_field *component` - if this would match the COBOL field, then it may be variable-length; but as far as I now see those are always fixed-length; so the issue is again "likely important, but likely big and belongs to another PR".

So... do you add key collations in cobc now and skip to set the collation if the field is numeric (potentially also skip if it is NATIONAL)? That should provide enough for the original issue.

> Yeah, they should always be of identical length in our context (and I made a quick check by adding an assert and running both NIST and the testsuite with no error). Anyway I prefered to handle the case anyways (the comparison function is rather "low-level" and does not really care about our use case).

The new indexed_compare _is_ our specialized function to _our_ case - and we cannot even support two different sizes as we don't have it when the specialized function is used.
Therefore: please adjust `indexed_compare()` to take only a single size, which then also removes the need to use `cob_min_int()`.

For the runtime warning - I think that should be an error, but we don't expect it to be seen by a user so remove the translation part and add a coverage exclusion, like

```c
			/* LCOV_EXCL_START */
			if (k++ == MAX_MODULE_ITERS) {	/* prevent endless loop in case of broken list */
				/* not translated as highly unexpected */
				cob_runtime_warning ("max module iterations exceeded, possible broken chain");
				break;
			}
			/* LCOV_EXCL_STOP */
```

GitMensch commented on cobc/cobc.c:
```diff
@@ -90,22 +90,23 @@ enum compile_level {
-#define	CB_FLAG_GETOPT_EBCDIC_TABLE    14
-#define	CB_FLAG_GETOPT_DEFAULT_COLSEQ  15
-#define	CB_FLAG_MEMORY_CHECK           16
+#define	CB_FLAG_GETOPT_STACK_SIZE            1
+#define	CB_FLAG_GETOPT_IF_CUTOFF             2
+#define	CB_FLAG_GETOPT_SIGN                  3
+#define	CB_FLAG_GETOPT_FOLD_COPY             4
+#define	CB_FLAG_GETOPT_FOLD_CALL             5
+#define	CB_FLAG_GETOPT_TTITLE                6
+#define	CB_FLAG_GETOPT_MAX_ERRORS            7
+#define	CB_FLAG_GETOPT_DUMP                  8
+#define	CB_FLAG_GETOPT_CALLFH                9
+#define	CB_FLAG_GETOPT_INTRINSICS           10
+#define	CB_FLAG_GETOPT_EC                   11
+#define	CB_FLAG_GETOPT_NO_EC                12
+#define	CB_FLAG_GETOPT_NO_DUMP              13
+#define	CB_FLAG_GETOPT_EBCDIC_TABLE         14
+#define	CB_FLAG_GETOPT_DEFAULT_COLSEQ       15
+#define	CB_FLAG_GETOPT_DEFAULT_FILE_COLSEQ  16
+#define	CB_FLAG_MEMORY_CHECK                17
```
Just wondering: do we want to make that an `enum` with `= 1` in the first entry, dropping all numbers in the comments of other places?
Ye, that is a refactoring that should be checked in separately up-front...

ddeclerck replied:
Yes indeed.

ddeclerck commented:
> > * either modifying the `bdb_bt_compare` function - but that would fix only for BDB...
> 
> that would be enough for now and is likely "the most portable" solution. But that creates a HUGE problem: if we do that, then the files need to be rebuilt :-/ Note that using the key type and use the appropriate COBOL sort function (if the field structure within ´->app_data`[if not done: store a complete cb_field structure in there], then you only need to duplicate that cob_field to then set the two separate fields' data) should directly work for all types. This is effectively what is done for`SEARCH`(and the internal comparison during`SORT`).
> 
> Otherwise we could have small-enough numbers be converted to big-endian binary and bigger ones to be converted to USAGE DISPLAY SIGN SEPARATE LEADING in `bdb_savekey`, that would also work - but again leads to all files needing a rebuild.

So, both solutions would imply a rebuilding of the files. What I like about the first one is that it uses backend's mechanisms intended for this purpose (i.e. custom key compare function) - but it has to be implemented specifically for each backend, and not all these backends expose such features. And what I like about the second one is that it is (I think ?) backend-agnostic.

> As this repacking/unpacking would be done on any key access, this is likely slower than the approach above.

I don't think we would ever have to translate the "encoded" keys backward, right ?

> As LMDB does not provide an external compare function, the second solution seems to be the only one that could be applied there

Isn't `mdb_set_compare` what we're looking for ?

> (but note that LMDB exists only in trunk and even then is marked as experimental and possibly will be dropped).

Even if slightly off-topic : I really like the idea of an LMDB backend. What would be the reasons for dropping it ? Not performing as expected ?

> Note: with ISAM there are key types that handle this (including IEE754), but we currently only use `CHARTYPE`, so the same problem seems to exists there. For this part I'll contact Ron.
> 
> > In any case, I believe this require some thinking before actually doing that, and I'm not sure this belongs to this PR (as it did not introduce that problem).
> 
> Hm. Please add a test with expected failure and a comment about the underlying problem.

I guess I'll do that.

> We could otherwise leave that general problem out of this PR, but if a key is purely numeric, there should be no "additional" conversion applied which makes the current scenario worse. That brings us to the good thing: if you implement key collation within the cobc site (you already have the codegen and use of that in libcob) then you can place a check there for a numeric type and just skip generating the collation for that.

Well, I think this is handled in `codegen.c: output_indexed_file_key_colseq` :
```
	int	type = cb_tree_type (key, cb_code_field (key));
	cb_tree	col = NULL;

	/* We only apply a collating sequence if the key is alphanumeric / display */
	if ((type & COB_TYPE_ALNUM) || (type == COB_TYPE_NUMERIC_DISPLAY)) {
		/* TODO: actually handle record keys (using file key for now) */
		col = f->collating_sequence;
	}
```
We agree that NUMERIC DISPLAY should be stored lexicographically, right ?

> > > As both can use the same comparison logic (just using a different set of collations), I think it's useful to test that.
> > 
> > Could you elaborate ?
> 
> `SEARCH`/`SORT` already compares "as cob_field" for any numeric type; if the same is applied to the INDEXED comparisons (you need a `cob_field` in ´app_data` for that), then everything "just works". But as we noted: this may be kept to a separate PR, _especially_ as we have to consider what to do with the existing files (likely we'd have to insert a marker into the file itself, likely by an embedded table, where we record "that's type 2, use different search algorithm"). This is definitely useful, but definitely "quite big" (and yes, should be kept separate).

Got it. As an alternative to an embedded "marker", how about an extra file for such metadata ?

> > > Testing variable length keys belong to this PR if previously possibly completely handled by BDB and now in our own search function.
> > 
> > I'm not sure to understand ; could you rephrase that ?
> 
> BDB compares keys. those are setup from `cob_field *component` - if this would match the COBOL field, then it may be variable-length; but as far as I now see those are always fixed-length; so the issue is again "likely important, but likely big and belongs to another PR".
> 
> So... do you add key collations in cobc now and skip to set the collation if the field is numeric (potentially also skip if it is NATIONAL)? That should provide enough for the original issue.

I think that's what my code does now ; except that it uses the file collation for every key - but I guess with all the things to do/fix we discussed here, I'd better implement complete support for (alphanumeric) key collations right now so we don't have to come back to it later.

> > Yeah, they should always be of identical length in our context (and I made a quick check by adding an assert and running both NIST and the testsuite with no error). Anyway I prefered to handle the case anyways (the comparison function is rather "low-level" and does not really care about our use case).
> 
> The new indexed_compare _is_ our specialized function to _our_ case - and we cannot even support two different sizes as we don't have it when the specialized function is used. Therefore: please adjust `indexed_compare()` to take only a single size, which then also removes the need to use `cob_min_int()`.

Alright, I'll change that.

> For the runtime warning - I think that should be an error, but we don't expect it to be seen by a user so remove the translation part and add a coverage exclusion, like
> 
> ```c
> 			/* LCOV_EXCL_START */
> 			if (k++ == MAX_MODULE_ITERS) {	/* prevent endless loop in case of broken list */
> 				/* not translated as highly unexpected */
> 				cob_runtime_warning ("max module iterations exceeded, possible broken chain");
> 				break;
> 			}
> 			/* LCOV_EXCL_STOP */
> ```

Will do.

ddeclerck commented:
Rabses, added key collations and a test with expected failure regarding the numeric keys sorting problem.
@GitMensch would it be sufficient for now ?

GitMensch commented:
while reviewing: I've just checked in the refactoring

GitMensch commented on cobc/codegen.c:
```diff
@@ -9271,6 +9271,40 @@ output_key_components (struct cb_file* f, struct cb_key_component* key_component
+{
+	cb_tree	key = ak ? ak->key : f->key;
+	cb_tree	key_col = ak ? ak->collating_sequence_key : f->collating_sequence_key;
+	int	type = cb_tree_type (key, cb_code_field (key));
```
please const that

GitMensch commented on cobc/codegen.c:
```diff
@@ -9271,6 +9271,40 @@ output_key_components (struct cb_file* f, struct cb_key_component* key_component
+	/* We only apply a collating sequence if the key is alphanumeric / display */
+	if ((type & COB_TYPE_ALNUM) || (type == COB_TYPE_NUMERIC_DISPLAY)) {
+		col = key_col ? key_col : f->collating_sequence;
+#if 0
+	/* TODO: this should be done for national, when available */
```
combine that to one line please - so it is clear why that is deactivated

GitMensch commented on cobc/codegen.c:
```diff
@@ -9271,6 +9271,40 @@ output_key_components (struct cb_file* f, struct cb_key_component* key_component
+	} else {
+		output ("(%s%s + %d)->collating_sequence = ", CB_PREFIX_KEYS, f->cname, idx);
+	}
+	if ((col != NULL) && CB_REFERENCE_P (col)) {
```
Not necessarily a change, more a question - is there a reason to wrap the first part in parens?

ddeclerck replied:
Ah, that's just my over-parenthesing OCD kicking-in. I'll drop them.

GitMensch commented on cobc/parser.y:
```diff
@@ -5752,7 +5741,7 @@ collating_sequence_clause:
 	current_file->collating_sequence = alphanumeric_collation;
 	current_file->collating_sequence_n = national_collation;
-	CB_PENDING ("FILE COLLATING SEQUENCE");
+	CB_UNFINISHED ("FILE COLLATING SEQUENCE");
```
please add an inline comment like `*/ only implemented for BDB /*`

GitMensch commented on libcob/fileio.c:
```diff
@@ -4150,6 +4170,30 @@ indexed_file_delete (cob_file *f, const char *filename)
+bdb_bt_compare(DB *db, const DBT *k1, const DBT *k2)
+{
+	const unsigned char *col = (unsigned char *)DBT_GET_APP_DATA(k1);
+	if (unlikely (col == NULL)) {
+		cob_runtime_error ("bdb_bt_compare was set but no collating sequence was stored in DBT");
+	}
+	if (unlikely (k1->size != k2->size)) {
```
for ease of later migrating to trunk: please drop `unlikely`, as this was removed; instead add LCOV exclusion comments around that, to not decrease coverage because of those checks

GitMensch commented on libcob/fileio.c:
```diff
@@ -757,6 +767,13 @@ bdb_savekey (cob_file *f, unsigned char *keyarea, unsigned char *record, int idx
 	return (int)f->keys[idx].field->size;
 }
 
+static void
```
should be forced-inlined

GitMensch reviewed (approved):
apart from the minor review comments and the rebase: that's beautiful - please: go, go, go

ddeclerck commented:
All changes accounted for - will merge to SVN if it's okay (I see the tag is already there).

GitMensch commented:
I trust your changes will be fine - as the test is already set: go on with the commit.

Thanks!

GitMensch commented:
@ddeclerck Thank you for working on this.
In my review I've missed that the new functions were defined globally - that broke the !WITH_DB builds, just fixed that in svn.
